Ignore non-semantic changes to objects

alexzielenski opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Ignores changes that only affect the timestamp in the managed fields. That can happen when someone merely updates a Quantity from 1024 to 1Ki or when accidentally removing a value that was a default. This change will reduce the total number of writes to etcd for usages which apply frequent changes to configuration with no semantic value.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #95460

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

apelisse commented:
/lgtm
/approve

binacs reviewed (approved):
/ok-to-test

apelisse commented:
/retest

apelisse commented:
/lgtm
/approve

/assign @liggitt

jpbetz commented:
/triage accepted

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,57 @@
+
+var equalities = func() conversion.Equalities {
+	e := equality.Semantic
+	e.AddFunc(ignoreManagedFieldsTimestamp)
```
this changes a process-wide equality 😱

sttts replied:
at least it touches the array below the slice, nearly as bad. Please do a deep copy of the slice first.

sttts replied:
Background:
```
e1 := equality.Semantic
e1.AddFunc(ignoreManagedFieldsTimestamp)
e2 := equality.Semantic
e2.AddFunc(someOtherEqualityFunc)
```
Which equality will be used in e1 and e2?

alexzielenski replied:
Nice catch, the map is now being shallow copied

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,57 @@
+//
+// The two objects are equal if they have the exact same values, and
+// their managed fieldsets are the same, except for the timestamp.
+func EqualObjects(new, old runtime.Object) bool {
```
this func is an anti-pattern, isn't it? Make equalities public and just call DeepEqual on it.

apelisse replied:
Actually, I don't think this is used, it can probably be removed

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,57 @@
+
+// DropEqualTransformer is a transformer that cancels the change if the
+// objects are equal.
+func DropEqualTransformer(_ context.Context, newObj, oldObj runtime.Object) (runtime.Object, error) {
```
IgnoreManagedFieldsTimestampsTransformer is a better name. These objects are not really equal. Reading DropEqualTransformer is misleading the reader.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,64 @@
+	oldObj runtime.Object,
+) (runtime.Object, error) {
+	if equalities.DeepEqual(newObj, oldObj) {
+		return oldObj, nil
```
to my knowledge, we've never made semantic equality a condition of persisting an update before... are we sure there are no objects that make distinctions that would be invisible to this check (`null` vs `[]`, etc)

cc @sttts @deads2k

apelisse replied:
As discussed with Jordan, we can make a strict equality with merely ignoring managed fields timestamp, which would be less intrusive.

sttts replied:
another clarification: these are internal objects or storage version objects? We do ignore updates of the later, we don't ignore updates of the former, or neither of request versions. It would break storage version migrator.

apelisse replied:
Ah, that's a good point. I'm curious why no test caught that?

None-the-less, rather than cancel the write (and make a strong decision), we could merely revert the updated timestamp in the managed fields. The etcd layer can then decide if it's worth writing the object.

alexzielenski replied:
I've adjusted the code to just revert the timestamp changes in managed fields, if that is all that has changed. (Rather than re-using the old object in its entirety).

apelisse replied:
yeah, I think conceptually this is doing a lot less than before (it only changes managedFields, not the rest of the object), which is much less risky.

apelisse replied:
To be honest, as we discussed with Alex, this was not preventing storage version migrator from updating the object since this was working on internal object, and not really by-passing the write, but keeping the same copy of the object.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,83 @@
+			return nil, err
+		}
+
+		newManagedFields := []metav1.ManagedFieldsEntry{}
```
you can probably allocate the right length here.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,83 @@
+			return reflect.DeepEqual(a, b)
+		},
+	)
+	for key, value := range equality.Semantic.Equalities {
```
Should we implement a "clone" method in the `Equalities` code?

alexzielenski replied:
Agreed

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,83 @@
+	if equalities.DeepEqual(newObj, oldObj) {
+		// Remove any changed timestamps, so that timestamp is not the only
+		// change seen by etcd.
+		//
```
Can you drop these empty comment lines (here and above?)

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,83 @@
+		//
+		oldAccessor, err := meta.Accessor(oldObj)
+		if err != nil {
+			return nil, err
```
Maybe we want to distinguish the error on the "new" accessor and the "old" accessor.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,86 @@
+	return e
+}()
+
+// IgnoreManagedFieldsTimestampsTransformer reverts any timestamp updates
```
not any

sttts commented on staging/src/k8s.io/apimachinery/pkg/conversion/deep_equal.go:
```diff
@@ -34,3 +34,16 @@ func EqualitiesOrDie(funcs ...interface{}) Equalities {
 	}
 	return e
 }
+
```
comment above Equalities is wrong

alexzielenski replied:
Hi Stefan, to my understanding the comment is correct. Is this not an implementation of a shallow copy for a map? Thanks

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,86 @@
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var equalities = func() conversion.Equalities {
```
it's not a good name. Be specific.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,86 @@
+			b.Time = nil
+
+			// Purposefully using reflect.DeepEqual, not semantic equality.
+			return reflect.DeepEqual(a, b)
```
why does it matter for ManagedFieldsEntry ?

sttts replied:
```golang
type ManagedFieldsEntry struct {
	// Manager is an identifier of the workflow managing these fields.
	Manager string `json:"manager,omitempty" protobuf:"bytes,1,opt,name=manager"`
	// Operation is the type of operation which lead to this ManagedFieldsEntry being created.
	// The only valid values for this field are 'Apply' and 'Update'.
	Operation ManagedFieldsOperationType `json:"operation,omitempty" protobuf:"bytes,2,opt,name=operation,casttype=ManagedFieldsOperationType"`
	// APIVersion defines the version of this resource that this field set
	// applies to. The format is "group/version" just like the top-level
	// APIVersion field. It is necessary to track the version of a field
	// set because it cannot be automatically converted.
	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,3,opt,name=apiVersion"`
	// Time is timestamp of when these fields were set. It should always be empty if Operation is 'Apply'
	// +optional
	Time *Time `json:"time,omitempty" protobuf:"bytes,4,opt,name=time"`

	// Fields is tombstoned to show why 5 is a reserved protobuf tag.
	//Fields *Fields `json:"fields,omitempty" protobuf:"bytes,5,opt,name=fields,casttype=Fields"`

	// FieldsType is the discriminator for the different fields format and version.
	// There is currently only one possible value: "FieldsV1"
	FieldsType string `json:"fieldsType,omitempty" protobuf:"bytes,6,opt,name=fieldsType"`
	// FieldsV1 holds the first JSON version format as described in the "FieldsV1" type.
	// +optional
	FieldsV1 *FieldsV1 `json:"fieldsV1,omitempty" protobuf:"bytes,7,opt,name=fieldsV1"`
}
```

apelisse replied:
> why does it matter for ManagedFieldsEntry ?

What do you mean by that?

sttts replied:
Why the semantic bit? For strings semantic and normal deep equal are the same.

sttts replied:
But ic that you compare the whole objects in https://github.com/kubernetes/kubernetes/pull/106388/files#diff-8604303e065a13f40ea019ec0b9a2fb733f44b5c6fbdd5b4b9238b442e94f527R58. For the rest it makes a difference 🤔

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,86 @@
+	newObj runtime.Object,
+	oldObj runtime.Object,
+) (runtime.Object, error) {
+	if equalities.DeepEqual(newObj, oldObj) {
```
does this mean that an array field changing from empty to undefined or back is not considered a change? @apelisse will remember this discussion. This needs a very careful analysis here.

apelisse replied:
Yeah, it's probably fine to do a non-semantic equality here, merely ignoring the timestamp. That'll also remove the need for the copy on `Equalities`.

apelisse replied:
Actually, the way `Equalities` work, we HAVE to use the semantic comparison (since it avoids the panic on the unexported fields). On the other hand, we're fine with the managedFields not being updated in that scenario since we already know that it means there are no changes to ownership.

alexzielenski replied:
The newest revision uses a version of deep equality which does not implicitly equate nil with empty slices/maps so I believe this conversation is resolved.

apelisse commented:
/retest

apelisse commented:
OK, we rediscussed and evaluated the exact semantics of what's going on here, I think we're good this time. Thanks.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/update.go:
```diff
@@ -182,6 +182,12 @@ func UpdateResource(r rest.Updater, scope *RequestScope, admit admission.Interfa
+		// like normalized fields, defaulted fields and other
+		// mutations.
+		transformers = append(transformers, fieldmanager.IgnoreManagedFieldsTimestampsTransformer)
+
```
I share @liggitt 's concerns that, while code-wise being clean now, this fixup logic still feels misplaced. Why does SSA even update the managed fields timestamp? Why do we need this fixup code being put in every update/patch code patch, even without SSA being involved?

To understand, this is the order:
1. SSA does its thing
2. mutating admission runs
3. defaulting runs
4. the resulting object migth be unchanged compared to oldObj, while step (1) saw a change?

sttts replied:
In other words, this transformer fights against an SSA patch that is continuesly trying to fight against some admission plugin or default. E.g. a mutating admission plugin sets `spec.privileged=false` for security policy reasons, but a GitOps system insists on `spec.privileged=true`. The result without this PR would be a hot loop with constantly updated managed fields timestamp.

sttts replied:
One maybe important detail for a questionable gut feeling: 

This logic uses semantic equality, which is uncommon elsewhere in the storage stack. What if we used normal, deep equality?

Then this would turn into:

a mutating admission plugin sets `spec.extraPermissions=[]` for security policy reasons, but a GitOps system insists on `spec.extraPermissions=undefined`. The result without this PR would be a hot loop with constantly updated managed fields timestamp.

sttts replied:
On the other hand, with this PR in place we would miss actual changes in etcd (where a mutating webhook set undefined to `[]`) where the managedField is not updated.

apelisse replied:
> Why does SSA even update the managed fields timestamp (my take below)? Why do we need this fixup code being put in every update/patch code patch, even without SSA being involved?

Because we don't want to track defaults/mutation changes as part of SSA so we run at the very beginning, but it also means that we don't know know what these changes are going to be and if they will undo a change.

> To understand, this is the order:
> 1. SSA does its thing
> 2. mutating admission runs
> 3. defaulting runs
> 4. the resulting object migth be unchanged compared to oldObj, while step (1) saw a change?

Yes, that's correct. Step 1 doesn't know what step 2 and 3 will do and specifically doesn't want to track these changes. But specifically when 2 and 3 undo what 1 saw, then we want to skip the extra etcd write (hence 4).

> The result without this PR would be a hot loop with constantly updated managed fields timestamp.

That's correct, that's the problem we're trying to solve here, and it's causing actual pain to users.

> This logic uses semantic equality, which is uncommon elsewhere in the storage stack. What if we used normal, deep equality

Note that this comparison is not by-passing any write to etcd. The only action it takes now is to by-pass the update to the timestamp in the managed fields.

We can not use deep equality because deep equality doesn't work for structures that have unexported fields.
In order to make it work, we had to copy the whole semantic deepequal (which doesn't do anything but provide custom comparison for kubernetes types that have unexported fields: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/equality/semantic.go#L29-L49). It also doesn't do anything specific for `undefined to []`.

> On the other hand, with this PR in place we would miss actual changes in etcd (where a mutating webhook sets undefined to []) where the managedField is not updated.

No, because we're not changing the object or preventing the object from going to etcd, we're just reverting the timestamp change. So if there is a change in the object, it's still going to go to etcd. If there is a change to managed fields other than the timestamp, it's still going to go to etcd.

sttts replied:
> It also doesn't do anything specific for undefined to [].
```
// An empty slice *is* equal to a nil slice for our purposes; same for maps.
```

> On the other hand, with this PR in place we would miss actual changes in etcd (where a mutating webhook sets undefined to []) where the managedField is not updated.

I meant that the object changes in etcd, but `managedFields` is not set. "miss" in that sense.

alexzielenski replied:
The latest revision includes a new function `Equalities.DeepEqualWithoutImplicitEqualities` which should not allow implicit equality for nil/empty slices/maps. I believe that resolves concerns about the empty slices/maps.

dims commented:
looks like this is not ready yet. (conversations not fully resolved)

apelisse commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -232,6 +245,14 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
+	return e.deepEqual(a1, a2, true)
+}
+
+func (e Equalities) DeepEqualWithoutImplicitEqualities(a1, a2 interface{}) bool {
```
I actually would have made this a property of the `Equalities` object if possible (might be tricky because of how we define the object), that avoids having to pass-in the parameter throughout the callstack, it gets copied automatically when you copy the semantic (we don't want to in that case, but I think it's reasonable in general).

alexzielenski replied:
Unfortunately your suspicion is correct that it is tricky because of how the object is defined.

apelisse commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -100,7 +100,7 @@ func makeUsefulPanic(v reflect.Value) {
 // comparisons that have already been seen, which allows short circuiting on
 // recursive types.
-func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, depth int) bool {
+func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, implicit bool, depth int) bool {
```
I can't see a comment changing, I'm surprised there's not a comment anywhere describing how that semantic comparison is done?

alexzielenski replied:
The latest version has a comment describing the new parameter

apelisse commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -150,17 +150,24 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
-		}
-		if v1.IsNil() || v1.Len() == 0 {
-			return true
+		if implicit {
```
I'm generally not a big big fan of "implicit" as a choice of word. Even in the name `DeepEqualWithoutImplicitEqualities`. These are implicit in a given context that nil lists and maps are implicitly equal to empty lists.

If we move forward with a member variable in the Equalities type, it should be false by default, so I'd go with:
`NilDifferentFromEmptyContainer` or something simpler if you can find one.

alexzielenski replied:
given that I cannot add `NilDifferentFromEmptyContainer` as a struct field, since the type is defined as a map as discussed in another part of the code, do you have any friendlier name recommendations for `DeepEqualWithoutImplicitEqualities` and the `implicit` argument?

alexzielenski replied:
it is renamed

sttts commented:
Sgtm. My concerns about semantics of the edge cases with nil and empty lists/maps are resolved with the latest updates. 

There have been questions about performance before when having this in every update code-path. Has those been addressed or have there been any benchmarks to mitigate those concerns?

alexzielenski commented:
I've ran this change, and the original code, quite a few times against the a WIP field validation benchmark `put-ignore`.

The results were surprising, indicating a possible small performance improvement.

```
name                                               old time/op    new time/op    delta
FieldValidation/Put/put-ignore-validation-16         2.32ms ± 8%    2.24ms ± 9%  -3.16%  (p=0.000 n=806+874)
FieldValidation/Put/put-ignore-validation-yaml-16    2.75ms ±11%    2.67ms ±12%  -2.85%  (p=0.000 n=742+734)
```

alexzielenski commented:
/cc @liggitt 

Took another benchmark with the new comparison taken in isolation.

The source code for this benchmark is here: https://gist.github.com/alexzielenski/8715898d08976e93eb59a42bb1d9f7b5

There are two benchmarks: for each of structured types and unstructured types. For each benchmark, a shallow object (v1.Endpoints) was tested (generated using fuzzer), and then a deep object (v1.Pod).

```
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/shallow
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/shallow-8         	    8595	    151961 ns/op	   32891 B/op	    1499 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/deep
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/deep-8            	     235	   5390087 ns/op	 1253875 B/op	   33811 allocs/op
```

```
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/shallow
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/shallow-8         	      46	  25047611 ns/op	 5193434 B/op	  188879 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/deep
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/deep-8            	     129	   8737373 ns/op	 1669834 B/op	   58405 allocs/op
```

I expect that the shallow object was slower for unstructured types due to the excessive amount of reflection done to iterate the map keys (very wide fan out). The deep object does not fan out nearly as much.

alexzielenski committed:
benchmark and metrics for new timestamp transformer comparison

add proper metrics

rename & improve documentation for path metric dimension

apelisse commented:
/retest

liggitt commented:
> There are two benchmarks: for each of structured types and unstructured types. For each benchmark, a shallow object (v1.Endpoints) was tested (generated using fuzzer), and then a deep object (v1.Pod).

huh... those are significantly more expensive than I hoped

apelisse commented on test/integration/apiserver/timestamp_transformer_test.go:
```diff
@@ -46,7 +47,12 @@ func BenchmarkIgnoreManagedFieldsTimestampTransformerStructured(b *testing.B) {
+	b.Logf("Shallow (Endpoints): %v bytes", len(bts))
+	bts, err = json.Marshal(pod)
+	require.NoError(b, err)
+	b.Logf("Deep (Pod): %v bytes", len(bts))
```
Can you do the same thing below?

alexzielenski commented:
> > There are two benchmarks: for each of structured types and unstructured types. For each benchmark, a shallow object (v1.Endpoints) was tested (generated using fuzzer), and then a deep object (v1.Pod).
> 
> huh... those are significantly more expensive than I hoped

I added a log to the benchmark to get a sense of what objects sort are being measured and ran again. The log prints the size of the object formatted as JSON in bytes.

```
    /Users/zielenski/go/src/k8s.io/kubernetes/test/integration/apiserver/timestamp_transformer_test.go:52: Shallow (Endpoints): 19839 bytes
    /Users/zielenski/go/src/k8s.io/kubernetes/test/integration/apiserver/timestamp_transformer_test.go:55: Deep (Pod): 170521 bytes
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/shallow
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/shallow-8         	    8953	    133800 ns/op	   45755 B/op	     999 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/deep
BenchmarkIgnoreManagedFieldsTimestampTransformerStructured/deep-8            	     618	   1868402 ns/op	  530219 B/op	   11178 allocs/op
```

This ran faster than the last time I ran the benchmark (I wonder if my computer was under load). The 2ms score for the `deep` benchmark is on a 170kb as json. Hopefully this provides some more context.

Here are results of the unstructured benchmarks along with the input size:

```
    /Users/zielenski/go/src/k8s.io/kubernetes/test/integration/apiserver/timestamp_transformer_test.go:116: Shallow (Endpoints): 1725411 bytes
    /Users/zielenski/go/src/k8s.io/kubernetes/test/integration/apiserver/timestamp_transformer_test.go:119: Deep (Pod): 610822 bytes
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/shallow
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/shallow-8         	      48	  25991266 ns/op	 5193476 B/op	  188879 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/deep
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructured/deep-8            	     122	  10176093 ns/op	 1826131 B/op	   63847 allocs/op
```

This also provides possibly a hint to the question of why the "shallow" unstructured benchmark results  seemed inexplicably faster than "deep": my "shallow" fuzzed object for this unstructured benchmark is not small (>1.5MB!). Also keep in mind that these "unstructured" tests populate most fields of the object, rather than a small subset.

apelisse commented:
OK So here's how I read this number, transforming them in human-readable:
- 133 µs for a 20k endpoint
- 1.9 ms for a 170k pod
- 26 ms for a 1.8M endpoint,
- 10ms for a 610k pods

I don't think that's that slow after-all?

liggitt commented:
as a floor to add to every request, it seems high. that's also an unfortunate number of allocs... I didn't realize a readonly deepequal check involved so many

lavalamp commented:
What are we blocked on here? I consider the bug this is fixing pretty critical.

(I also consider that number of allocs to be slightly ridiculous for comparison checks, but I haven't looked at the code here)

liggitt commented:
> What are we blocked on here? I consider the bug this is fixing pretty critical.

uncertainty whether that's a reasonable number of allocs / compute to add to every write request

lavalamp commented:
@liggitt what are numbers you could live with? Just to get this out of limbo.

liggitt commented:
> @liggitt what are numbers you could live with?

I don't know what our numbers are currently at, but the allocs especially seem egregious.

Did we get scaleability runs with/without this? We're actually in the middle of recovering from a scale regression today (https://github.com/kubernetes/kubernetes/issues/109146), so we'd need clean signal before getting data on this.

If this merges and we discover it causes problems, what's our recourse? Wait for a fix/backport/patch release?

alexzielenski commented:
@liggitt @lavalamp You guys are correct to be suspicious of the number of allocations. I attached a debugger to the benchmark and I have found that the branch the test is taking the "transformation" code path, rather than simply the "comparison" code path:

https://github.com/kubernetes/kubernetes/pull/106388/files#diff-8604303e065a13f40ea019ec0b9a2fb733f44b5c6fbdd5b4b9238b442e94f527R58

This could very likely explain the high number of deep copies.

This is surprising since the benchmark tests two identical objects, [one being a deep copy of the other](https://github.com/kubernetes/kubernetes/pull/106388/files#diff-3e2d935ca4990d4678efcf482d4de1f9fe647109f846b9663fab39a7ae38d5fbR111). 

I'll post back in a bit with either updated benchmark results if it is a problem with the benchmark, or more interesting results if there is a bug with the deepCopy/comparison.

alexzielenski commented:
EDIT: Deleted earlier comment

lavalamp commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -100,7 +100,8 @@ func makeUsefulPanic(v reflect.Value) {
 // recursive types.
-func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, depth int) bool {
+// equateNilAndEmpty controls whether empty maps/slices are equivalent to nil
+func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, equateNilAndEmpty bool, depth int) bool {
```
Visited could be a `map[visit]struct{}`, that will make it smaller but probably not reduce the number of allocations

lavalamp replied:
Why does this function have both depth and a visited map? Is depth actually used?

alexzielenski replied:
It appears as if `depth` is only ever `+1`'d for recursive calls

lavalamp replied:
Why does the `visit` type include the type and not just the pointers?

liggitt replied:
looks like there have been optimizations applied since we forked this in 2017:

* https://github.com/golang/go/commit/c4f9fa1f8ee623558ed40d4d4cd3c616697cc77b
* https://github.com/golang/go/commit/a0c409cbc82ea9999a03fa0bfe6d9a8953e780e0

also, bugfixes:

* https://github.com/golang/go/commit/4dc11ae26b3a092bbcd1b4d0ae8cdaeb1cf1cef9
* https://github.com/golang/go/commit/d872bbcfec29f2585ed8144fba6262d4228619cf
* https://github.com/golang/go/commit/009d71409821a6ac4f1b32aaae2c856c20a29f92

liggitt replied:
it looks like we also have two copies of this... third_party/forked/golang/reflect/deep_equal.go and staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go

outside the scope of this PR, but we should have at most one, and should be paying attention to what lands upstream to pick up bug/perf fixes

for this PR, can we at least avoid diverging our two copies so that we can collapse onto one in the future?

alexzielenski replied:
We we know if `third_party/forked/golang/reflect` is used anywhere? A search of k8s codebase via `rg` makes it seem unused?

```
src/k8s.io/kubernetes
❯ rg third_party/forked/golang/reflect 
vendor/modules.txt
1419:k8s.io/apimachinery/third_party/forked/golang/reflect
1927:k8s.io/code-generator/third_party/forked/golang/reflect

staging/src/k8s.io/apimachinery/pkg/conversion/deep_equal.go
20:	"k8s.io/apimachinery/third_party/forked/golang/reflect"

staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/parser.go
33:	customreflect "k8s.io/code-generator/third_party/forked/golang/reflect"
```

(the customreflect package at the end is a fork of only `types.go`: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/code-generator/third_party/forked/golang/reflect)

lavalamp replied:
delete it and see if everything still compiles...

liggitt replied:
Done in https://github.com/kubernetes/kubernetes/pull/109244, just pending master reopen

alexzielenski replied:
I had already submitted a PR for its removal. It is being removed in 1.25 milestone here...

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,83 @@
+	newObj runtime.Object,
+	oldObj runtime.Object,
+) (runtime.Object, error) {
+	if ignoreTimestampEqualities.DeepEqualWithNilDifferentFromEmpty(newObj, oldObj) {
```
Do you need to compare the entire objects? or is it OK to just compare the managed fields?

liggitt replied:
the bug is that managedFields changed timestamps when nothing else changed in the rest of the object

liggitt replied:
I had thought about an initial fast(er)/cheap(er) check to see if managedFields was identical except for timestamps, ahead of the full object comparison, but:
* that will likely be true in the case of high-volume controller-driven status updates touching the same fields, which is where most of our traffic comes from
* getting managedFields out via an accessor adds its own allocations for custom resources: https://github.com/kubernetes/kubernetes/blob/64c5ed380441b4015301cf64a44a698400b03d8b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L459-L479

lavalamp replied:
OK, you've read the first thought out of my mind, but I have more:

Has generation been set at this point? If so, you can short-circuit a spec check by looking at that.

We can compare everything *except* for managed fields -- if that is different, then there is no need to compare or modify managed fields.

I believe we must already be doing a deep equal somewhere before writing; if we can at this point somehow pass along a "yes this is really modified" flag with the object, we could skip that later check, making this more neutral.

liggitt replied:
> Has generation been set at this point? If so, you can short-circuit a spec check by looking at that.

Generation happens in PrepareForUpdate, after this point

> I believe we must already be doing a deep equal somewhere before writing; if we can at this point somehow pass along a "yes this is really modified" flag with the object, we could skip that later check, making this more neutral.

The only equality check I know if in the pipeline is the byte-comparison where etcd storage is considering whether the bytes we're about to write are equal. That's too late for this logic (and would include the managedField differences we're trying to avoid here)

lavalamp replied:
I still think pre-checking managed fields for changes that are only timestamps and nothing else before doing the full check may have value.

Since this is only triggered by a small number of cases (nonstandard resource.Quantity, what else?), we could restrict this to a hardcoded list of endpoints that could possibly trigger this. (e.g. probably no status update needs to go through this code.)

If we REALLY want to go down into the weeds, that byte comparison could possibly be made a proto-tag-aware comparison and notice when the only thing different is this particular field (even easier for json). We can depend on well-formatted data at that point so potentially that would be very cheap.

The other alternative is just to debug the current code and figure out why allocations are so high, of course.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,104 @@
+	// This condition ensures the managed fields are always compared first. If
+	//	this check fails, the if statement will short circuit. If the check
+	// 	succeeds the slow path is taken which compares entire objects.
+	if ignoreTimestampEqualities.DeepEqualWithNilDifferentFromEmpty(oldAccessor, accessor) &&
```
Shouldn't we just compare the managed fields like so?
```suggestion
		if ignoreTimestampEqualities.DeepEqualWithNilDifferentFromEmpty(oldAccessor.GetManagedFields(), accessor.GetManagedFields()) &&
```

alexzielenski replied:
fixed

apelisse commented:
I think that's clearly an improvement and I don't think, even for controller regularly updating the status, that this would fall in the worst case scenario?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,104 @@
+	// This condition ensures the managed fields are always compared first. If
+	//	this check fails, the if statement will short circuit. If the check
+	// 	succeeds the slow path is taken which compares entire objects.
+	if ignoreTimestampEqualities.DeepEqualWithNilDifferentFromEmpty(oldAccessor.GetManagedFields(), accessor.GetManagedFields()) &&
```
We extract them again on line 92/93 -- is this expensive enough (or is the compiler dumb enough) that it's worth pre-getting old and new managed fields to avoid redundant work?

Also can we get updated perf numbers for both does and doesn't match cases?

liggitt replied:
> is this expensive enough

it's expensive for unstructured objects: https://github.com/kubernetes/kubernetes/blob/7380fc735aca591325ae1fabf8dab194b40367de/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L459-L479

alexzielenski replied:
Addressed in latest commits. Also posted new benchmark results. Marking this thread as resolved.

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,83 @@
+
+		for idx, value := range oldManagedFields {
+			// Deep copy each managed field entry in case the original is modified
+			value.DeepCopyInto(&newManagedFields[idx])
```
You are correct. Given that the old and new managedfields are equal except for their timestamps, it is safe and sufficient to simply replace new managed field's time with that of their older counterpart.

fixed in new commit. marking as resolved

alexzielenski commented:
The transformer has been modified to add short circuiting logic, as well as other comments. 

I've also straightened out the benchmark code & nomenclature so that it is clearer. Here are new results comparing short circuiting logic for both large and small objects:

```
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/apiserver
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkIgnoreManagedFieldsTimestampTransformerStructuredShortCircuit/Large-8           7468573	       147.7 ns/op	      48 B/op	       2 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerStructuredShortCircuit/Small-8         	 8039496	       147.0 ns/op	      48 B/op	       2 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerStructuredWorstCase/Large-8            	      61	  20580477 ns/op	 5123732 B/op	  163281 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerStructuredWorstCase/Small-8            	     211	   5346769 ns/op	 1961422 B/op	   38134 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructuredShortCircuit/Large-8        	    7712	    154329 ns/op	   88219 B/op	     949 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructuredShortCircuit/Small-8        	   68330	     18849 ns/op	   10305 B/op	     138 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructuredWorstCase/Large-8           	      70	  16271151 ns/op	 3318406 B/op	  145322 allocs/op
BenchmarkIgnoreManagedFieldsTimestampTransformerUntructuredWorstCase/Small-8           	     140	   8677537 ns/op	 1547008 B/op	   68951 allocs/op
PASS
```

As expected, short circuiting logic is much more performant. Large/Small object size makes no difference for short circuiting structured objects since length can easily be compared. For unstructured there is a small time difference due to how managed fields must be created for the comparison.

lavalamp commented:
I think we can live with that if the worst case is rare -- to figure that out, can you figure out a way to run the e2e tests, counting how often the short circuit is helpful?

alexzielenski commented:
I instrumented the code to log three variables as well as average time for each path taken. I then ran the e2e tests with the following command (which refers to my kind checkout) `FOCUS="apply|aggregator|chunking|webhook|watch|protocol|discovery" ../../sigs.k8s.io/kind/hack/ci/e2e-k8s.sh`

Here are the results from running those e2e tests:

| | Equal Objects | Unequal Objects - Short Circuit | Unequal Objects - Full Comparison |
|-----|---------------|--------------------------------|------------------------------------|
| Time | 63.800ms |  5.838ms | 37.588ms |
| Occurrences | 5 | 519 | 2468 |

It is unclear how representative this is of real data. At some point in the test the distribution of occurrences was 5 (equal) 446 (short circuit) 1275 (full comparison). So it may be possible a specific test is stressing that code path. Unfortunately at least to me it is not clear from the logs which test was running when it happened.

lavalamp commented:
Thanks! There almost certainly is a test that tests us not writing unchanged stuff, but I'm not sure it would affect the numbers that much.

I'm not worried about 30us extra, but if that's the relative ratio then I am worried about all the allocations. :(

DWSR commented:
Hey, this bug just cost my company ~$3k in intra-region transfer in AWS (Istio reacted to a constantly "updating" Service by pushing out new Envoy configs). Any chance this can get merged in and backported to 1.22 soon-ish?

alexzielenski commented:
Summarizing offline discussions with @lavalamp, @apelisse @liggitt :

Path forward for this change includes:

1. Instrumenting the change with metrics to see distribution of code paths (equal vs unequal comparison, fast vs slow path)
2. Adding a flag to easily disable the bugfix in the case it becomes too expensive.

1. I have updated the PR with a draft of the metrics instrumentation to track allocations and latency for 3 different code paths of the code.
2. For the feature flag I am unsure what form would be appropriate to take? In discussions I've heard feature flag is not appropriate? Is there an example of an alternative being used elsewhere in source so I do not write something ad-hoc?

apelisse commented:
I don't have a strong opinion about the escape hatch. I have a feeling that this won't cause any specific problem, and if it does we can roll-forward or backward without the change. I think the metrics would help us detect if such a problem arise.
Can we move forward with that?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -238,6 +238,26 @@ var (
 
+	requestTimestampComparisonDuration = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name:           "apiserver_request_timestamp_comparison_time",
```
@logicalhan We're not necessarily planning on having these two metrics for a very long time, do you have any recommendations?

logicalhan replied:
use alpha as the metric stability level? and annotate it in the HELP description?

apelisse replied:
Sounds good, thanks!

liggitt commented:
are you expecting to backport to release branches? *if* we do backport, I'd lean towards:
* making the escape hatch available so a re-release isn't required if a problem is found
* making the behavior change opt-in in backports

alexzielenski committed:
add envar to disable non semantic updates feature

enabled by default. can easily be changed in backports

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -238,6 +238,26 @@ var (
+			Buckets:        []float64{0.0001, 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1.0, 5.0},
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"path"},
```
we synced offline about this

summary:
"path" here referred not to request path, but code path.


has been renamed to "code_path" and commented

alexzielenski committed:
revert timestamp updates to object if non-managed fields do not change

add short-circuiting logic for long comaprison

replace timestamps rather than doing a full managed fields deepcopy

add guard

liggitt commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -178,22 +186,28 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
+		} else {
+			if v1.IsNil() != v2.IsNil() {
+				return false
+			}
```
for parity with the `equateNilAndEmpty` branch, should this have a `return true` case optimizing `v1.IsNil() && v2.IsNil()` and `v1.Len() == 0 && v2.Len() == 0`?

alexzielenski replied:
added

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -660,7 +660,7 @@ func (p *patcher) patchResource(ctx context.Context, scope *RequestScope) (runti
 
 	wasCreated := false
-	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyAdmission, dedupOwnerReferencesTransformer)
+	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyAdmission, dedupOwnerReferencesTransformer, fieldmanager.IgnoreManagedFieldsTimestampsTransformer)
```
do we need to check if `scope.FieldManager` is non-nil here like we do in the update handler?

alexzielenski replied:
added

alexzielenski committed:
optimize nil and empty case for parity with other branch

alexzielenski committed:
guard usage of timestamp transformer under fieldManager non nil

not strictly necessary for correctness, but it is not needed unless SSA is enabled

liggitt commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -150,17 +151,24 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
+		} else {
+			if v1.IsNil() != v2.IsNil() {
+				return false
+			}
```
same question about the both nil and both empty optimization case here

alexzielenski replied:
I feel bad that I missed this. Thanks

liggitt commented:
two last nits, then lgtm

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,180 @@
+	// changed timestamp; and are otherwise equal. Return early if there are no
+	// changed timestamps.
+	allTimesUnchanged := true
+	for i, e := range newManagedFields {
```
This is intended as a quick short circuit if none if the times changed, since the operation is meaningless without that.

There is a call to DeepEqual on just the managed fields below this line which would filter out any cases where a re-ordering occurred.

alexzielenski committed:
optimize nil and empty also for slices

brings to parity with maps

alexzielenski committed:
use more apt name for flag

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:
```diff
@@ -0,0 +1,180 @@
+)
+
+func determineIgnoreNonSemanticUpdatesEnabled() bool {
+	if ignoreNonSemanticUpdatesString, exists := os.LookupEnv("KUBE_APISERVER_IGNORE_NON_SEMANTIC_UPDATES"); exists {
```
changed now to `KUBE_APISERVER_AVOID_NOOP_SSA_TIMESTAMP_UPDATES`

liggitt commented:
With the metrics and escape hatch, this looks reasonably scoped to me. I'd like a second approver ack on the final shape (@lavalamp and @sttts seem like good candidates since they were both involved in reviewing this)

alexzielenski commented:
/retest

lavalamp commented:
/lgtm
/approve
